Add tests to make sure shared subgraph works fine

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35256 Add tests to make sure shared subgraph works fine**

Address the subset associated with overlapping subgraphs in https://github.com/pytorch/pytorch/issues/34367 .

albanD committed:
Add tests to make sure shared subgraph works fine

[ghstack-poisoned]

pritamdamania87 commented on test/test_autograd.py:
```diff
@@ -4014,6 +4015,86 @@ def backward(ctx, grad):
+
+    def test_engine_shared_graph_threading(self):
+        # This test tries a "worst case scenario" to make sure that we don't call a single Node concurrently
+        # from multiple threads while multiple backward passes are running.
```
I'm assuming this test passes since we just have a single CPU thread. Does this test make sense with multi-threaded autograd?

albanD replied:
This would be part of the changed behaviors by https://github.com/pytorch/pytorch/pull/33157 indeed.
I'll just remove it as it does not make sense to put it in just to remote it in the other PR.

pritamdamania87 commented on test/test_autograd.py:
```diff
@@ -4014,6 +4015,86 @@ def backward(ctx, grad):
+            @staticmethod
+            def backward(ctx, grad_output):
+                # Make sure nothing else enters here before we're done
+                if MyFunction.in_use:
```
Shouldn't we do this with a lock/semaphor where we try to grab the lock/semaphore in the beginning of the function and throw an exception if we cannot?

albanD replied:
Would it behave differently? I am not really a specialist of these modules in python so I used a simple global flag as we don't need any actual locking. You think I should change it?

albanD replied:
Not relevant since I'm removing this test.

wanchaol commented on test/test_autograd.py:
```diff
@@ -4014,6 +4015,86 @@ def backward(ctx, grad):
+
+        def run_backward():
+            out = model_end(middle)
+            grads = torch.autograd.grad(out.sum(), model_shared.parameters(), retain_graph=True)
```
The expected behavior of this is that in the current master engine with no parallelism on CPU, this will never raise, but with parallel autograd engine this should raise right? Because `time.sleep(0.2)` release the GIL (we rely on GIL for PyNode to ensure thread safety rather than locking ourselves), and other tasks can enter the same Node, which will raise exception right?

albanD replied:
Yes, removing this to avoid merge issues with the new engine.

albanD committed:
Update on "Add tests to make sure shared subgraph works fine"


Address the subset associated with overlapping subgraphs in https://github.com/pytorch/pytorch/issues/34367 .

[ghstack-poisoned]
